File: examples/webhooks/fastapi/post_call.py
"""
ElevenLabs post-call webhook — FastAPI implementation.

Receives conversation data after a call ends, persists it to a database,
and sends notifications for escalated or unsuccessful calls.
"""

from __future__ import annotations

import asyncio
import logging
import os
from datetime import datetime, timezone
from typing import Any

import httpx
from fastapi import APIRouter, HTTPException, Request
from pydantic import BaseModel, Field

logger = logging.getLogger(__name__)

# ---------------------------------------------------------------------------
# Configuration
# ---------------------------------------------------------------------------

DATABASE_URL: str = os.environ["DATABASE_URL"]
NOTIFICATION_WEBHOOK_URL: str | None = os.environ.get("NOTIFICATION_WEBHOOK_URL")

# ---------------------------------------------------------------------------
# Models
# ---------------------------------------------------------------------------


class DataCollectionField(BaseModel):
    name: str
    value: str
    description: str | None = None


class EvaluationResult(BaseModel):
    result: str
    rationale: str


class CallAnalysis(BaseModel):
    call_successful: str | None = None
    transcript_summary: str | None = None
    data_collection_results: dict[str, DataCollectionField] = Field(
        default_factory=dict
    )
    evaluation_criteria_results: dict[str, EvaluationResult] = Field(
        default_factory=dict
    )


class CallCost(BaseModel):
    total_cost_usd: float


class PostCallEvent(BaseModel):
    conversation_id: str
    agent_id: str
    call_sid: str | None = None
    caller_id: str | None = None
    call_duration_secs: float
    status: str  # done | failed | busy | no-answer
    call_cost: CallCost | None = None
    metadata: dict[str, str] = Field(default_factory=dict)
    analysis: CallAnalysis | None = None


class PostCallWebhookPayload(BaseModel):
    type: str
    event: PostCallEvent


# ---------------------------------------------------------------------------
# Database
# ---------------------------------------------------------------------------


async def insert_call_record(event: PostCallEvent) -> None:
    """Insert a call log row into the database.

    Uses asyncpg directly. Swap in SQLAlchemy, Tortoise, or any async ORM
    as needed.
    """
    import asyncpg  # type: ignore[import-untyped]

    data_collection = {
        k: v.value
        for k, v in (event.analysis.data_collection_results.items() if event.analysis else {})
        if v.value
    }

    conn: asyncpg.Connection = await asyncpg.connect(DATABASE_URL)
    try:
        await conn.execute(
            """
            INSERT INTO call_logs (
                conversation_id, agent_id, caller_id, duration_secs,
                status, cost_usd, summary, call_successful,
                data_collection, metadata, created_at
            ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9::jsonb, $10::jsonb, $11)
            """,
            event.conversation_id,
            event.agent_id,
            event.caller_id,
            event.call_duration_secs,
            event.status,
            event.call_cost.total_cost_usd if event.call_cost else None,
            event.analysis.transcript_summary if event.analysis else None,
            event.analysis.call_successful if event.analysis else None,
            data_collection,
            event.metadata,
            datetime.now(timezone.utc).isoformat(),
        )
    finally:
        await conn.close()


# ---------------------------------------------------------------------------
# Notifications
# ---------------------------------------------------------------------------


def _should_notify(
    event: PostCallEvent,
) -> tuple[bool, str]:
    analysis = event.analysis

    if analysis and analysis.call_successful == "false":
        return True, "Call was unsuccessful"

    if analysis:
        for criterion, result in analysis.evaluation_criteria_results.items():
            if "escalat" in criterion.lower() and result.result == "true":
                return True, f"Escalation detected: {criterion}"
            if "sentiment" in criterion.lower() and result.result == "negative":
                return True, f"Negative sentiment: {result.rationale}"

    if event.call_duration_secs > 600:
        return (
            True,
            f"Extended call duration: {round(event.call_duration_secs / 60)} minutes",
        )

    return False, ""


async def send_notification(event: PostCallEvent, reason: str) -> None:
    if not NOTIFICATION_WEBHOOK_URL:
        logger.warning("NOTIFICATION_WEBHOOK_URL not set, skipping: %s", reason)
        return

    summary = (
        event.analysis.transcript_summary
        if event.analysis and event.analysis.transcript_summary
        else "No summary available."
    )
    duration_min = round(event.call_duration_secs / 60)

    payload = {
        "channel": "#call-alerts",
        "text": f"Call alert: {reason}",
        "blocks": [
            {
                "type": "header",
                "text": {"type": "plain_text", "text": f"Call Alert: {reason}"},
            },
            {
                "type": "section",
                "fields": [
                    {
                        "type": "mrkdwn",
                        "text": f"*Conversation:*\n{event.conversation_id}",
                    },
                    {
                        "type": "mrkdwn",
                        "text": f"*Caller:*\n{event.caller_id or 'Unknown'}",
                    },
                    {"type": "mrkdwn", "text": f"*Duration:*\n{duration_min} min"},
                    {"type": "mrkdwn", "text": f"*Status:*\n{event.status}"},
                ],
            },
            {
                "type": "section",
                "text": {"type": "mrkdwn", "text": f"*Summary:*\n{summary}"},
            },
        ],
    }

    async with httpx.AsyncClient() as client:
        resp = await client.post(
            NOTIFICATION_WEBHOOK_URL,
            json=payload,
            timeout=5.0,
        )
        if resp.status_code >= 400:
            logger.error("Notification failed: %d %s", resp.status_code, resp.text)


# ---------------------------------------------------------------------------
# Router
# ---------------------------------------------------------------------------

router = APIRouter()


@router.post("/webhooks/elevenlabs/post-call")
async def post_call_webhook(request: Request) -> dict[str, str]:
    try:
        body = await request.json()
        payload = PostCallWebhookPayload(**body)
    except Exception as exc:
        raise HTTPException(status_code=400, detail=f"Invalid payload: {exc}")

    if payload.type != "post_call":
        raise HTTPException(
            status_code=422,
            detail=f"Unexpected webhook type: {payload.type}",
        )

    event = payload.event

    # Run DB insert and notification concurrently.
    db_task = asyncio.create_task(insert_call_record(event))

    notify_task: asyncio.Task[None] | None = None
    should_notify, reason = _should_notify(event)
    if should_notify:
        notify_task = asyncio.create_task(send_notification(event, reason))

    results = await asyncio.gather(
        db_task,
        *([] if notify_task is None else [notify_task]),
        return_exceptions=True,
    )

    # The DB insert is at index 0.
    if isinstance(results[0], BaseException):
        logger.error("Failed to insert call record: %s", results[0])
        raise HTTPException(status_code=500, detail="Failed to persist call data")

    # Notification failure at index 1 (if present) is non-critical.
    if len(results) > 1 and isinstance(results[1], BaseException):
        logger.error("Notification failed: %s", results[1])

    return {"status": "ok", "conversation_id": event.conversation_id}

File: examples/webhooks/express/post-call.ts
import { Request, Response, Router } from "express";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

interface DataCollectionField {
  name: string;
  value: string;
  description?: string;
}

interface PostCallWebhookPayload {
  type: "post_call";
  event: {
    conversation_id: string;
    agent_id: string;
    call_sid?: string;
    caller_id?: string;
    call_duration_secs: number;
    status: "done" | "failed" | "busy" | "no-answer";
    call_cost?: {
      total_cost_usd: number;
    };
    metadata?: Record<string, string>;
    analysis?: {
      call_successful?: "true" | "false" | "unknown";
      transcript_summary?: string;
      data_collection_results?: Record<string, DataCollectionField>;
      evaluation_criteria_results?: Record<
        string,
        { result: string; rationale: string }
      >;
    };
  };
}

interface CallRecord {
  conversation_id: string;
  agent_id: string;
  caller_id: string | null;
  duration_secs: number;
  status: string;
  cost_usd: number | null;
  summary: string | null;
  call_successful: string | null;
  data_collection: Record<string, string>;
  metadata: Record<string, string>;
  created_at: string;
}

// ---------------------------------------------------------------------------
// Configuration
// ---------------------------------------------------------------------------

const DATABASE_URL = process.env.DATABASE_URL!;
const NOTIFICATION_WEBHOOK_URL = process.env.NOTIFICATION_WEBHOOK_URL;

// ---------------------------------------------------------------------------
// Database
// ---------------------------------------------------------------------------

async function insertCallRecord(record: CallRecord): Promise<void> {
  const { default: postgres } = await import("postgres");
  const sql = postgres(DATABASE_URL, { max: 1, idle_timeout: 5 });

  try {
    await sql`
      INSERT INTO call_logs (
        conversation_id, agent_id, caller_id, duration_secs, status,
        cost_usd, summary, call_successful, data_collection, metadata, created_at
      ) VALUES (
        ${record.conversation_id}, ${record.agent_id}, ${record.caller_id},
        ${record.duration_secs}, ${record.status}, ${record.cost_usd},
        ${record.summary}, ${record.call_successful},
        ${JSON.stringify(record.data_collection)},
        ${JSON.stringify(record.metadata)}, ${record.created_at}
      )
    `;
  } finally {
    await sql.end();
  }
}

// ---------------------------------------------------------------------------
// Notifications
// ---------------------------------------------------------------------------

function shouldNotify(payload: PostCallWebhookPayload): {
  notify: boolean;
  reason: string;
} {
  const analysis = payload.event.analysis;

  if (analysis?.call_successful === "false") {
    return { notify: true, reason: "Call was unsuccessful" };
  }

  const evaluations = analysis?.evaluation_criteria_results ?? {};
  for (const [criterion, result] of Object.entries(evaluations)) {
    if (
      criterion.toLowerCase().includes("escalat") &&
      result.result === "true"
    ) {
      return { notify: true, reason: `Escalation detected: ${criterion}` };
    }
    if (
      criterion.toLowerCase().includes("sentiment") &&
      result.result === "negative"
    ) {
      return {
        notify: true,
        reason: `Negative sentiment: ${result.rationale}`,
      };
    }
  }

  if (payload.event.call_duration_secs > 600) {
    return {
      notify: true,
      reason: `Extended call duration: ${Math.round(payload.event.call_duration_secs / 60)} minutes`,
    };
  }

  return { notify: false, reason: "" };
}

async function sendNotification(
  payload: PostCallWebhookPayload,
  reason: string,
): Promise<void> {
  if (!NOTIFICATION_WEBHOOK_URL) {
    console.warn("NOTIFICATION_WEBHOOK_URL not set, skipping:", reason);
    return;
  }

  const event = payload.event;
  const summary =
    event.analysis?.transcript_summary ?? "No summary available.";
  const durationMin = Math.round(event.call_duration_secs / 60);

  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), 5_000);

  try {
    const res = await fetch(NOTIFICATION_WEBHOOK_URL, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        channel: "#call-alerts",
        text: `Call alert: ${reason}`,
        blocks: [
          {
            type: "header",
            text: { type: "plain_text", text: `Call Alert: ${reason}` },
          },
          {
            type: "section",
            fields: [
              {
                type: "mrkdwn",
                text: `*Conversation:*\n${event.conversation_id}`,
              },
              {
                type: "mrkdwn",
                text: `*Caller:*\n${event.caller_id ?? "Unknown"}`,
              },
              { type: "mrkdwn", text: `*Duration:*\n${durationMin} min` },
              { type: "mrkdwn", text: `*Status:*\n${event.status}` },
            ],
          },
          {
            type: "section",
            text: { type: "mrkdwn", text: `*Summary:*\n${summary}` },
          },
        ],
      }),
      signal: controller.signal,
    });

    if (!res.ok) {
      console.error(`Notification failed: ${res.status} ${res.statusText}`);
    }
  } finally {
    clearTimeout(timer);
  }
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function extractDataCollection(
  results?: Record<string, DataCollectionField>,
): Record<string, string> {
  if (!results) return {};
  const out: Record<string, string> = {};
  for (const [key, field] of Object.entries(results)) {
    if (field.value) out[key] = field.value;
  }
  return out;
}

// ---------------------------------------------------------------------------
// Router
// ---------------------------------------------------------------------------

const router = Router();

router.post(
  "/webhooks/elevenlabs/post-call",
  async (req: Request, res: Response): Promise<void> => {
    const payload = req.body as PostCallWebhookPayload;

    if (payload.type !== "post_call") {
      res.status(422).json({ error: `Unexpected type: ${payload.type}` });
      return;
    }

    const event = payload.event;

    const record: CallRecord = {
      conversation_id: event.conversation_id,
      agent_id: event.agent_id,
      caller_id: event.caller_id ?? null,
      duration_secs: event.call_duration_secs,
      status: event.status,
      cost_usd: event.call_cost?.total_cost_usd ?? null,
      summary: event.analysis?.transcript_summary ?? null,
      call_successful: event.analysis?.call_successful ?? null,
      data_collection: extractDataCollection(
        event.analysis?.data_collection_results,
      ),
      metadata: event.metadata ?? {},
      created_at: new Date().toISOString(),
    };

    const [dbResult, notifyResult] = await Promise.allSettled([
      insertCallRecord(record),
      (async () => {
        const { notify, reason } = shouldNotify(payload);
        if (notify) await sendNotification(payload, reason);
      })(),
    ]);

    if (dbResult.status === "rejected") {
      console.error("Failed to insert call record:", dbResult.reason);
      res.status(500).json({ error: "Failed to persist call data" });
      return;
    }

    if (notifyResult.status === "rejected") {
      console.error("Notification failed:", notifyResult.reason);
    }

    res.json({ status: "ok", conversation_id: event.conversation_id });
  },
);

export default router;
